File: VocaDbModel/DataContracts/Venues/ArchivedVenueVersionDetailsContract.cs
#nullable disable

using System.Linq;
using System.Runtime.Serialization;
using VocaDb.Model.DataContracts.Users;
using VocaDb.Model.DataContracts.Versioning;
using VocaDb.Model.Domain.Security;
using VocaDb.Model.Domain.Venues;

namespace VocaDb.Model.DataContracts.Venues
{
	public class ArchivedVenueVersionDetailsContract
	{
#nullable enable
		public ArchivedVenueVersionDetailsContract(ArchivedVenueVersion archived, ArchivedVenueVersion? comparedVersion, IUserPermissionContext permissionContext, IUserIconFactory userIconFactory)
		{
			ParamIs.NotNull(() => archived);

			ArchivedVersion = new ArchivedVenueVersionContract(archived, userIconFactory);
			ComparedVersion = comparedVersion != null ? new ArchivedVenueVersionContract(comparedVersion, userIconFactory) : null;
			Venue = new VenueContract(archived.Entry, permissionContext.LanguagePreference);
			Name = Venue.Name;

			ComparableVersions = archived.Entry.ArchivedVersionsManager
				.GetPreviousVersions(archived, permissionContext)
				.Select(a => ArchivedObjectVersionWithFieldsContract.Create(a, userIconFactory, a.Diff.ChangedFields.Value, a.CommonEditEvent))
				.ToArray();

			Versions = ComparedVenueContract.Create(archived, comparedVersion);

			ComparedVersionId = Versions.SecondId;
		}
#nullable disable

		public ArchivedVenueVersionContract ArchivedVersion { get; init; }

		public ArchivedObjectVersionContract[] ComparableVersions { get; init; }

		public ArchivedObjectVersionContract ComparedVersion { get; init; }

		public int ComparedVersionId { get; init; }

		public bool Hidden => ArchivedVersion.Hidden || (ComparedVersion != null && ComparedVersion.Hidden);

		public string Name { get; init; }

		public VenueContract Venue { get; init; }

		public ComparedVenueContract Versions { get; init; }
	}
}

File: VocaDbWeb/Scripts/DataContracts/Venue/ArchivedVenueVersionDetailsContract.ts
import { ArchivedVenueContract } from '@/DataContracts/Venue/ArchivedVenueContract';
import { VenueForApiContract } from '@/DataContracts/Venue/VenueForApiContract';
import { ArchivedVersionContract } from '@/DataContracts/Versioning/ArchivedVersionContract';
import { ComparedVersionsContract } from '@/DataContracts/Versioning/ComparedVersionsContract';

// Corresponds to the ArchivedVenueVersionDetailsForApiContract record class in C#.
export interface ArchivedVenueVersionDetailsContract {
	venue: VenueForApiContract;
	archivedVersion: ArchivedVersionContract;
	comparableVersions: ArchivedVersionContract[];
	comparedVersion?: ArchivedVersionContract;
	name: string;
	versions: ComparedVersionsContract<ArchivedVenueContract>;
}
